File: tools/stub-compile/gesture_mods_equiv.js
// JS side of the gesture-mods parity check (see gesture_mods_equiv.cpp).
// Dumps the REAL applyGestureModsLane() from the built bundle for the same cases,
// in the same "name|count|curve,dur,value|…" format, so run.sh can diff them.
//   node gesture_mods_equiv.js <path/to/lib/pardalote.js>
const fs = require('fs');
const bundlePath = process.argv[2];
const src = fs.readFileSync(bundlePath, 'utf8');
const { applyGestureModsLane } = new Function('global', 'document',
    src + '\nreturn { applyGestureModsLane };')(global, undefined);

const C = { linear: 0, easeIn: 1, easeOut: 2, easeInOut: 3, back: 4 };
const cid = (c) => (typeof c === 'number' ? c : (C[c] ?? 0));

// Relative "nod" (by) and its absolute (to) twin — mirror the C++ cases.
const nod = [
    { by:  190, dur: 260, curve: 'easeInOut' }, { by: -190, dur: 260, curve: 'easeInOut' },
    { by:  130, dur: 260, curve: 'easeInOut' }, { by: -130, dur: 260, curve: 'back' },
];
const nodAbs = [
    { to: 1858, dur: 260, curve: 'easeInOut' }, { to: 2048, dur: 260, curve: 'easeInOut' },
    { to: 1918, dur: 260, curve: 'easeInOut' }, { to: 2048, dur: 260, curve: 'back' },
];
const val = (s) => (s.by !== undefined ? s.by : (s.value !== undefined ? s.value : s.to));

function run(name, segs, mods) {
    const out = applyGestureModsLane(segs, { scale: mods[0], speed: mods[1], crop: [mods[2], mods[3]] }, () => {});
    let line = `${name}|${out.length}`;
    for (const s of out) line += `|${cid(s.curve)},${Math.max(1, Math.round(s.dur))},${val(s)}`;
    console.log(line);
}

run('identity',  nod,    [1.0, 1.0, 0.0, 1.0]);
run('scale.5',   nod,    [0.5, 1.0, 0.0, 1.0]);
run('speed2',    nod,    [1.0, 2.0, 0.0, 1.0]);
run('speed.5',   nod,    [1.0, 0.5, 0.0, 1.0]);
run('crop.2-.8', nod,    [1.0, 1.0, 0.2, 0.8]);
run('combo',     nod,    [0.7, 1.5, 0.1, 0.9]);
run('abs-combo', nodAbs, [0.5, 2.0, 0.25, 0.75]);

File: tools/stub-compile/gesture_mods_equiv.cpp
// ==============================================================
// Board vs JS parity for gesture mods (scale / speed / crop).
//
// Compiles the REAL board transform — pardaloteApplyModsInPlace() from
// internal/gesture.h (via the Arduino stub) — and dumps its output for a set of
// cases. gesture_mods_equiv.js dumps the REAL JS applyGestureModsLane() from the
// bundle for the SAME cases; run.sh diffs the two. No hand-replicated logic, so
// the JS↔board maths (order, rounding, pardaloteEase↔curveShape) can't drift.
//
// Build:  g++ -std=gnu++17 -I stubs -I <src>/internal gesture_mods_equiv.cpp
// ==============================================================
#include <cstdio>
#include "gesture.h"

static PardaloteGestureMod M(float sc, float sp, float cf, float ct) {
    PardaloteGestureMod m; m.scale = sc; m.speed = sp; m.cropFrom = cf; m.cropTo = ct; return m;
}

static void run(const char* name, const PardaloteSeg* src, uint8_t n, uint8_t flags, PardaloteGestureMod m) {
    PardaloteSeg buf[16];
    for (uint8_t i = 0; i < n; i++) buf[i] = src[i];
    uint8_t k = pardaloteApplyModsInPlace(buf, n, flags, m);
    printf("%s|%u", name, k);
    for (uint8_t i = 0; i < k; i++) printf("|%u,%u,%d", buf[i].curve, buf[i].dur, buf[i].value);
    printf("\n");
}

int main() {
    const uint8_t IO = CURVE_EASE_IN_OUT, BK = CURVE_BACK;
    // A relative "nod": down/up/down/up-settle (values are deltas).
    PardaloteSeg nod[4]    = {{IO,260,190},{IO,260,-190},{IO,260,130},{BK,260,-130}};
    // The same shape as absolute targets.
    PardaloteSeg nodAbs[4] = {{IO,260,1858},{IO,260,2048},{IO,260,1918},{BK,260,2048}};

    run("identity",  nod, 4, 0, M(1.0f, 1.0f, 0.0f, 1.0f));
    run("scale.5",   nod, 4, 0, M(0.5f, 1.0f, 0.0f, 1.0f));
    run("speed2",    nod, 4, 0, M(1.0f, 2.0f, 0.0f, 1.0f));
    run("speed.5",   nod, 4, 0, M(1.0f, 0.5f, 0.0f, 1.0f));
    run("crop.2-.8", nod, 4, 0, M(1.0f, 1.0f, 0.2f, 0.8f));
    run("combo",     nod, 4, 0, M(0.7f, 1.5f, 0.1f, 0.9f));
    run("abs-combo", nodAbs, 4, GESTURE_FLAG_ABSOLUTE, M(0.5f, 2.0f, 0.25f, 0.75f));
    return 0;
}
